<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>9. CpG_to_gene.py &mdash; CpGtools documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b76e3c8a" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=c97855ec" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=b3ba4146"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="10. beta_PCA.py" href="beta_PCA.html" />
    <link rel="prev" title="8. CpG_logo.py" href="CpG_logo.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >

          
          
          <a href="../index.html" class="icon icon-home">
            CpGtools
              <img src="../_static/CpGtools_logo2.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                2.0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">CpGtools Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">1. Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">2. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../change_log.html">3. CpGtools Release history</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Data Format</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../input_format.html">1. Input file and data format</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Test and Demo Datasets</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../dataset.html">1. Pre-compiled datasets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Information</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="CpG_aggregation.html">1. CpG_aggregation.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_anno_position.html">2. CpG_anno_position.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_anno_probe.html">3. CpG_anno_probe.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_density_gene_centered.html">4. CpG_density_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_chrom.html">5. CpG_distrb_chrom.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_gene_centered.html">6. CpG_distrb_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_region.html">7. CpG_distrb_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_logo.html">8. CpG_logo.py</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">9. CpG_to_gene.py</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#description">9.1. Description</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id1">9.2. Description</a></li>
<li class="toctree-l2"><a class="reference internal" href="#notes">9.3. Notes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#options">9.4. Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#input-files-examples">9.5. Input files (examples)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#command">9.6. Command</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-files">9.7. Output files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="beta_PCA.html">10. beta_PCA.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_UMAP.html">11. beta_UMAP.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_jitter_plot.html">12. beta_jitter_plot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_m_conversion.html">13. beta_m_conversion.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_profile_gene_centered.html">14. beta_profile_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_profile_region.html">15. beta_profile_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_stacked_barplot.html">16. beta_stacked_barplot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_stats.html">17. beta_stats.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_tSNE.html">18. beta_tSNE.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_topN.html">19. beta_topN.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_trichotmize.html">20. beta_trichotmize.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_Bayes.html">21. dmc_Bayes.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_bb.html">22. dmc_bb.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_fisher.html">23. dmc_fisher.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_glm.html">24. dmc_glm.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_logit.html">25. dmc_logit.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_nonparametric.html">26. dmc_nonparametric.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_ttest.html">27. dmc_ttest.py</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Compare Differential Analysis Tools</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../compare_diff_tools.html">1. Compare Differential CpG Analysis Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../compare_diff_tools.html#p-value-distributions">2. P-value distributions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Liscense</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../liscense.html">LICENSE</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reference.html">Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: white" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CpGtools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">9. </span>CpG_to_gene.py</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/demo/CpG_to_gene.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="cpg-to-gene-py">
<h1><span class="section-number">9. </span>CpG_to_gene.py<a class="headerlink" href="#cpg-to-gene-py" title="Permalink to this heading"></a></h1>
<section id="description">
<h2><span class="section-number">9.1. </span>Description<a class="headerlink" href="#description" title="Permalink to this heading"></a></h2>
<p>This program annotates CpGs by assigning them to their putative target genes. It follows the
“Basal plus extension rules” used by <a class="reference external" href="http://great.stanford.edu/public/html/">GREAT</a>.</p>
<p>Basal regulatory domain is a user-defined genomic region around the TSS (transcription
start site). By default, from TSS upstream 5 Kb to TSS downstream 1 Kb is considered as
the gene’s basal regulatory domain. When defining a gene’s basal regulatory domain, the
other nearby genes are ignored (which means different genes’ basal regulatory domain can
be overlapped.)</p>
<p>Extended regulatory domain is a genomic region that is further extended from basal
regulatory domain in both directions to the nearest gene’s basal regulatory domain but
no more than the maximum extension (specified by ‘-e’, default - 1000 kb) in one
direction.      In other words, the “extension” stops when it reaches other genes’ “basal
regulatory domain” or the extension limit, whichever comes first.</p>
<p>Basal regulatory domain and Extended regulatory domain are illustrated in below diagram.</p>
<a class="reference internal image-reference" href="../_images/gene_domain.png"><img alt="../_images/gene_domain.png" src="../_images/gene_domain.png" style="width: 600.0px; height: 200.0px;" /></a>
<p><strong>Notes</strong></p>
<ul class="simple">
<li><p>Which genes are assigned to a particular CpG largely depends on gene annotation. A
“conservative” gene model (such as Refseq curated protein-coding genes) is recommended.</p></li>
<li><p>In the refgene file, multiple isoforms should be merged into a single gene.</p></li>
</ul>
</section>
<section id="id1">
<h2><span class="section-number">9.2. </span>Description<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h2>
<p>This program annotates CpGs by assigning them to their putative target genes. Follows the
“Basel plus extension” rules used by GREAT(<a class="reference external" href="http://great.stanford.edu/public/html/index.php">http://great.stanford.edu/public/html/index.php</a>)</p>
<blockquote>
<div><ul class="simple">
<li><p>Basal regulatory domain:
is a user-defined genomic region around the TSS (transcription start site). By default,
from TSS upstream 5kb to TSS downstream 1Kb is considered as the gene’s <em>basal regulatory
domain</em>. When defining a gene’s “basal regulatory domain”, the other nearby genes will be
ignored (which means different genes’ basal regulatory domains can be overlapped.)</p></li>
<li><p>Extended regulatory domain:
The gene regulatory domain is extended in both directions to the nearest gene’s “basal
regulatory domain” but no more than the maximum extension (default = 1000 kb) in one
direction.</p></li>
</ul>
</div></blockquote>
</section>
<section id="notes">
<h2><span class="section-number">9.3. </span>Notes<a class="headerlink" href="#notes" title="Permalink to this heading"></a></h2>
<blockquote>
<div><ol class="arabic simple">
<li><p>Which genes are assigned to a particular CpG largely depends on gene annotation. A
“conservative” gene model (such as Refseq curated protein coding genes) is recommended.</p></li>
<li><p>In the gene model, multiple isoforms should be merged into a single gene.</p></li>
</ol>
</div></blockquote>
</section>
<section id="options">
<h2><span class="section-number">9.4. </span>Options<a class="headerlink" href="#options" title="Permalink to this heading"></a></h2>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-i <var>INPUT_FILE</var></span>, <span class="option">--input-file=<var>INPUT_FILE</var></span></kbd></dt>
<dd><p>BED3+ file specifying the C position. BED3+ file could
be a regular text file or compressed file (.gz, .bz2).
[required]</p>
</dd>
<dt><kbd><span class="option">-r <var>GENE_FILE</var></span>, <span class="option">--refgene=<var>GENE_FILE</var></span></kbd></dt>
<dd><p>Reference gene model in BED12 format
(<a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">https://genome.ucsc.edu/FAQ/FAQformat.html#format1</a>).
“One gene one transcript” is recommended. Since most
genes have multiple transcripts; one can collapse
multiple transcripts of the same gene into a single
super transcript or select the canonical transcript.</p>
</dd>
<dt><kbd><span class="option">-u <var>BASAL_UP_SIZE</var></span>, <span class="option">--basal-up=<var>BASAL_UP_SIZE</var></span></kbd></dt>
<dd><p>Size of extension to upstream of TSS (used to define
gene’s “basal regulatory domain”). default=5000 (bp)</p>
</dd>
<dt><kbd><span class="option">-d <var>BASAL_DOWN_SIZE</var></span>, <span class="option">--basal-down=<var>BASAL_DOWN_SIZE</var></span></kbd></dt>
<dd><p>Size of extension to downstream of TSS (used to define
gene’s basal regulatory domain). default=1000 (bp)</p>
</dd>
<dt><kbd><span class="option">-e <var>EXTENSION_SIZE</var></span>, <span class="option">--extension=<var>EXTENSION_SIZE</var></span></kbd></dt>
<dd><p>Size of extension to both up- and down-stream of TSS
(used to define gene’s “extended regulatory domain”).
default=1000000 (bp)</p>
</dd>
<dt><kbd><span class="option">-o <var>OUT_FILE</var></span>, <span class="option">--output=<var>OUT_FILE</var></span></kbd></dt>
<dd><p>Prefix of the output file. Two additional columns will
be appended to the original BED file with the last
column indicating “genes whose extended regulatory
domain are overlapped with the CpG”, the 2nd last
column indicating “genes whose basal regulatory domain
are overlapped with the CpG”. [required]</p>
</dd>
</dl>
</dd>
</dl>
</section>
<section id="input-files-examples">
<h2><span class="section-number">9.5. </span>Input files (examples)<a class="headerlink" href="#input-files-examples" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/test/850K_probe.hg19.bed3.gz">850K_probe.hg19.bed3.gz</a></p></li>
<li><p><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/refgene/hg19.RefSeq.union.bed.gz">hg19.RefSeq.union.bed.gz</a></p></li>
</ul>
</section>
<section id="command">
<h2><span class="section-number">9.6. </span>Command<a class="headerlink" href="#command" title="Permalink to this heading"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CpG_to_gene.py -i  850K_probe.hg19.bed3.gz -r hg19.RefSeq.union.bed.gz -o output
</pre></div>
</div>
</section>
<section id="output-files">
<h2><span class="section-number">9.7. </span>Output files<a class="headerlink" href="#output-files" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>output.associated_genes.txt</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head output.associated_genes.txt

#The last column contains genes whose extended regulatory domain are overlapped with the CpG
#The 2nd last column contains genes whose basal regulatory domain are overlapped with the CpG
#&quot;//&quot; indicates no genes are found
chr1   10524   10525   DDX11L1 //
chr1   10847   10848   DDX11L1 //
chr1   10849   10850   DDX11L1 //
chr1   15864   15865   //      MIR6859-1;DDX11L1
chr1   18826   18827   MIR6859-1       //
chr1   29406   29407   WASH7P;MIR1302-2        //
chr1   29424   29425   WASH7P;MIR1302-2        //
...
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="CpG_logo.html" class="btn btn-neutral float-left" title="8. CpG_logo.py" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="beta_PCA.html" class="btn btn-neutral float-right" title="10. beta_PCA.py" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Liguo Wang.
      <span class="lastupdated">Last updated on Sep 05, 2024.
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>