

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>3. CpG_anno_probe.py &mdash; CpGtools documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=c97855ec" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9a2dae69"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="4. CpG_density_gene_centered.py" href="CpG_density_gene_centered.html" />
    <link rel="prev" title="2. CpG_anno_position.py" href="CpG_anno_position.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >

          
          
          <a href="../index.html" class="icon icon-home">
            CpGtools
              <img src="../_static/CpGtools_logo3.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">CpGtools Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">1. Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">2. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../change_log.html">3. CpGtools Release history</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Data Format</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../input_format.html">1. Input file and data format</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Test and Demo Datasets</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../dataset.html">1. Pre-compiled datasets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Information</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="CpG_aggregation.html">1. CpG_aggregation.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_anno_position.html">2. CpG_anno_position.py</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">3. CpG_anno_probe.py</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#description">3.1. Description</a></li>
<li class="toctree-l2"><a class="reference internal" href="#notes">3.2. Notes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#options">3.3. Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#input-files-examples">3.4. Input files (examples)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#command">3.5. Command</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-files">3.6. Output files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CpG_density_gene_centered.html">4. CpG_density_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_chrom.html">5. CpG_distrb_chrom.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_gene_centered.html">6. CpG_distrb_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_region.html">7. CpG_distrb_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_logo.html">8. CpG_logo.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_to_gene.html">9. CpG_to_gene.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_PCA.html">10. beta_PCA.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_UMAP.html">11. beta_UMAP.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_jitter_plot.html">12. beta_jitter_plot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_m_conversion.html">13. beta_m_conversion.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_profile_gene_centered.html">14. beta_profile_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_profile_region.html">15. beta_profile_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_remove_batch_effects.html">16. beta_remove_batch_effects.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_selectNBest.html">17. beta_selectNBest.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_stacked_barplot.html">18. beta_stacked_barplot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_stats.html">19. beta_stats.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_tSNE.html">20. beta_tSNE.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_topN.html">21. beta_topN.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_trichotmize.html">22. beta_trichotmize.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_Bayes.html">23. dmc_Bayes.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_bb.html">24. dmc_bb.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_fisher.html">25. dmc_fisher.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_glm.html">26. dmc_glm.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_logit.html">27. dmc_logit.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_nonparametric.html">28. dmc_nonparametric.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_ttest.html">29. dmc_ttest.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="predict_sex.html">30. predict_sex.py</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Compare Differential Analysis Tools</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../compare_diff_tools.html">1. Compare Differential CpG Analysis Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../compare_diff_tools.html#p-value-distributions">2. P-value distributions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Liscense</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../liscense.html">LICENSE</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reference.html">Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: white" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CpGtools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">3. </span>CpG_anno_probe.py</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/demo/CpG_anno_probe.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="cpg-anno-probe-py">
<h1><span class="section-number">3. </span>CpG_anno_probe.py<a class="headerlink" href="#cpg-anno-probe-py" title="Link to this heading"></a></h1>
<section id="description">
<h2><span class="section-number">3.1. </span>Description<a class="headerlink" href="#description" title="Link to this heading"></a></h2>
<p>This program adds comprehensive annotation information to each 450K/850K array probe ID.
It will add 17 columns to the original input data file. These 17 columns include
(from left to right):</p>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>Header Name</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-even"><td><p>hg19_pos</p></td>
<td><p>The genomic position of the CpG on human genome assembly <a class="reference external" href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.13/">hg19 (or
GRCh37)</a></p></td>
</tr>
<tr class="row-odd"><td><p>hg38_pos</p></td>
<td><p>The genomic position of the CpG on human genome assembly <a class="reference external" href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.26/">hg38 (or
GRCh38)</a>.</p></td>
</tr>
<tr class="row-even"><td><p>strand</p></td>
<td><p>Strand of the CpG. Value - “R” (reverse strand) or “F” (forward strand).</p></td>
</tr>
<tr class="row-odd"><td><p>geneSymbol</p></td>
<td><p>Genes the CpG has been assigned to. “N/A” indicates no genes were found.
This is retrieved from the Illumina <a class="reference external" href="https://support.illumina.com/downloads/infinium-methylationepic-v1-0-product-files.html">MethylationEpic v1.0 B4</a> manifest file.</p></td>
</tr>
<tr class="row-even"><td><p>CpGisland</p></td>
<td><p>The CpG island (CGI) that overlaps with this CpG. “N/A” indicates no
CGIs were found.</p></td>
</tr>
<tr class="row-odd"><td><p>with_450K</p></td>
<td><p>Boolean indicating whether this CpG probe is also included in 450K.
“0” - No, “1”- Yes.</p></td>
</tr>
<tr class="row-even"><td><p>SNP_ID</p></td>
<td><p>SNPs (rsID) that are close to this CpG. Multiple SNPs are separated
by “;”. “N/A” indicates no SNPs were found.</p></td>
</tr>
<tr class="row-odd"><td><p>SNP_distance</p></td>
<td><p>The nucleotide distances between SNPs and the CpG.</p></td>
</tr>
<tr class="row-even"><td><p>SNP_MAF</p></td>
<td><p>The <a class="reference external" href="https://en.wikipedia.org/wiki/Minor_allele_frequency">minor allele frequencies (MAF)</a> of SNPs.</p></td>
</tr>
<tr class="row-odd"><td><p>Cross_Reactive</p></td>
<td><p>Boolean (“0” - No, “1”- Yes) indicating whether this CpG could be
affected by cross-hybridization or underlying genetic variation as
reported by this <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1066-1">paper</a>.</p></td>
</tr>
<tr class="row-even"><td><p>ENCODE_TF_ChIP</p></td>
<td><p>Transcription factor (TF) binding sites identified from ChIP-seq
experiments performed by the <a class="reference external" href="https://www.encodeproject.org/">ENCODE</a>
project. Peaks from 1264 experiments representing 338 transcription
factors in 130 cell types are combined (N = 10,560,472).
BED format file was downloaded from the <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgTables">UCSC Tabel Browser</a>, and a detailed description
is provided <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgTrackUi?hgsid-732007223_QUJBO5BMeBu3R7xczOAWQ0UV9A1f&amp;c-chr9&amp;g-encRegTfbsClustered">here</a>.</p></td>
</tr>
<tr class="row-odd"><td><p>ENCODE_DNaseI</p></td>
<td><p>DNase I hypersensitivity sites identified from ENCODE <a class="reference external" href="https://en.wikipedia.org/wiki/DNase-Seq">DNase-seq</a> experiments. Peaks from
125 cell types are combined (N - 1,867,665). BED format file was
downloaded from the <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgTables">UCSC Table Browser</a>, and a detailed description
is provided <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgTrackUi?hgsid-732007223_QUJBO5BMeBu3R7xczOAWQ0UV9A1f&amp;c-chr9&amp;g-wgEncodeRegDnaseClustered">here</a>.</p></td>
</tr>
<tr class="row-even"><td><p>ENCODE_H3K27ac_ChIP</p></td>
<td><p>H3K27ac peaks identified from ENCODE histone ChIP-seq experiments. Peaks
from 11 cell types (GM12878, H1-hESC, HMEC, HSMM, HUVEC, HeLaS3, HepG2,
K562, Monocytes-CD14+_RO01746, NHEK, NHLF) are combined (N = 665,650)</p></td>
</tr>
<tr class="row-odd"><td><p>ENCODE_H3K4me1_ChIP</p></td>
<td><p>H3K4me1 peaks identified from ENCODE histone ChIP-seq experiments. Peaks
from 11 cell types (GM12878, H1-hESC, HMEC, HSMM, HUVEC, HeLaS3, HepG2,
K562, Monocytes-CD14+_RO01746, NHEK, NHLF) are combined (N = 1,435,550)</p></td>
</tr>
<tr class="row-even"><td><p>ENCODE_H3K4me3_ChIP</p></td>
<td><p>H3K4me3 peaks identified from ENCODE histone ChIP-seq experiments. Peaks
from 11 cell types (GM12878, H1-hESC, HMEC, HSMM, HUVEC, HeLaS3, HepG2,
K562, Monocytes-CD14+_RO01746, NHEK, NHLF) are combined (N = 525,824)</p></td>
</tr>
<tr class="row-odd"><td><p>ENCODE_chromHMM</p></td>
<td><p>Chromatin State Segmentation by <a class="reference external" href="https://www.nature.com/articles/nmeth.1906">chromHMM</a> from ENCODE. Chromatin states across 9 cell types
(GM12878,  H1-hESC, K562, HepG2, HUVEC, HMEC, HSMM, NHEK, NHLF) were
learned by computationally by integrating 9 factors (CTCF, H3K27ac,
H3K27me3, H3K36me3, H3K4me1, H3K4me2, H3K4me3, H3K9ac, H4K20me1 )
plus input. A total of 15 states were identified, include: State-1
(Active Promoter), state-2 (Weak Promoter), state-3 (Inactive/poised
Promoter), state-4 and 5 (Strong enhancer), state-6 and 7
(Weak/poised enhancer), state-8 (insulator), state-9 (Transcriptional
transition), state-10 (Transcriptional elongation), state-11 (Weak
transcribed), state-12 (Polycomb-repressed), state-13 (Heterochromatin or
low signal), state-14 and 15 (Repetitive/Copy Number Variation).
Orignal chromatin state BED file was downloaded from <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgTables">UCSC Table Browser</a>, and detailed description
is provided <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgTrackUi?hgsid-732007223_QUJBO5BMeBu3R7xczOAWQ0UV9A1f&amp;c-chr9&amp;g-wgEncodeBroadHmm">here</a>.</p></td>
</tr>
<tr class="row-even"><td><p>FANTOM_enhancer</p></td>
<td><p>PHANTOM5 human enhancers downloaded from <a class="reference external" href="http://fantom.gsc.riken.jp/5/datafiles/latest/extra/Enhancers/human_permissive_enhancers_phase_1_and_2_expression_tpm_matrix.txt.gz">here</a>.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="notes">
<h2><span class="section-number">3.2. </span>Notes<a class="headerlink" href="#notes" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>For peaks identified from ENCODE ChIP-seq and DNase-seq (ENCODE_TF_ChIP, ENCODE_H3K27ac_ChIP,
ENCODE_H3K4me1_ChIP, ENCODE_H3K4me3_ChIP, and ENCODE_DNaseI), we require the probe  must be
located in the 100 bp window centered on the <strong>middle</strong> of the peak.</p></li>
</ul>
</section>
<section id="options">
<h2><span class="section-number">3.3. </span>Options<a class="headerlink" href="#options" title="Link to this heading"></a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-i <var>INPUT_FILE</var></span>, <span class="option">--input_file=<var>INPUT_FILE</var></span></kbd></dt>
<dd><p>Input data file (Tab-separated) with a certain column
containing 450K/850K array CpG IDs. This file can be
a regular text file or compressed file (.gz, .bz2).</p>
</dd>
<dt><kbd><span class="option">-a <var>ANNO_FILE</var></span>, <span class="option">--annotation=<var>ANNO_FILE</var></span></kbd></dt>
<dd><p>Annotation file. This file can be a regular text file
or compressed file (.gz, .bz2).</p>
</dd>
<dt><kbd><span class="option">-o <var>OUT_FILE</var></span>, <span class="option">--output=<var>OUT_FILE</var></span></kbd></dt>
<dd><p>Prefix of the output file.</p>
</dd>
<dt><kbd><span class="option">-p <var>PROBE_COL</var></span>, <span class="option">--probe_column=<var>PROBE_COL</var></span></kbd></dt>
<dd><p>The number specifying which column contains probe IDs.
Note: the column index starts with 0. default-0.</p>
</dd>
<dt><kbd><span class="option">-l</span>, <span class="option">--header</span></kbd></dt>
<dd><p>Input data file has a header row.</p>
</dd>
</dl>
</div></blockquote>
</section>
<section id="input-files-examples">
<h2><span class="section-number">3.4. </span>Input files (examples)<a class="headerlink" href="#input-files-examples" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/test/test_01.hg19.bed6">test_01.hg19.bed6</a></p></li>
<li><p><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/MethylationEPIC_CpGtools.tsv.gz">MethylationEPIC_CpGtools.tsv.gz</a></p></li>
</ul>
</section>
<section id="command">
<h2><span class="section-number">3.5. </span>Command<a class="headerlink" href="#command" title="Link to this heading"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># probe IDs are located in the 4th column (-p 3)

$CpG_anno_probe.py -p 3 -l -a MethylationEPIC_CpGtools.tsv -i test_01.hg19.bed6 -o output

or (take gzipped files as input)

$CpG_anno_probe.py -p 3 -l -a MethylationEPIC_CpGtools.tsv.gz -i test_01.hg19.bed6.gz -o output

@ 2019-06-28 09:12:41: Read annotation file &quot;../epic/MethylationEPIC_CpGtools.tsv&quot; ...
@ 2019-06-28 09:12:52: Add annotation information to &quot;test_01.hg19.bed6&quot; ...
</pre></div>
</div>
</section>
<section id="output-files">
<h2><span class="section-number">3.6. </span>Output files<a class="headerlink" href="#output-files" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>output.anno.txt</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="CpG_anno_position.html" class="btn btn-neutral float-left" title="2. CpG_anno_position.py" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="CpG_density_gene_centered.html" class="btn btn-neutral float-right" title="4. CpG_density_gene_centered.py" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Liguo Wang.
      <span class="lastupdated">Last updated on Oct 07, 2025.
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>