

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>2. CpG_anno_position.py &mdash; CpGtools documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="3. CpG_anno_probe.py" href="CpG_anno_probe.html" />
    <link rel="prev" title="1. CpG_aggregation.py" href="CpG_aggregation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> CpGtools
          

          
            
            <img src="../_static/CpGtools_logo2.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">CpGtools Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">1. Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">2. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../change_log.html">3. CpGtools Release history</a></li>
</ul>
<p class="caption"><span class="caption-text">Data Format</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../input_format.html">1. Input file and data format</a></li>
</ul>
<p class="caption"><span class="caption-text">Test and Demo Datasets</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../dataset.html">1. Pre-compiled datasets</a></li>
</ul>
<p class="caption"><span class="caption-text">Usage Information</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="CpG_aggregation.html">1. CpG_aggregation.py</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">2. CpG_anno_position.py</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#description">2.1. Description</a></li>
<li class="toctree-l2"><a class="reference internal" href="#notes">2.2. Notes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pre-computed-datasets">2.3. Pre-computed datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#options">2.4. Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#input-files-examples">2.5. Input files (examples)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#command">2.6. Command</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-files">2.7. Output files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CpG_anno_probe.html">3. CpG_anno_probe.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_density_gene_centered.html">4. CpG_density_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_chrom.html">5. CpG_distrb_chrom.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_gene_centered.html">6. CpG_distrb_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_distrb_region.html">7. CpG_distrb_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_logo.html">8. CpG_logo.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="CpG_to_gene.html">9. CpG_to_gene.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_PCA.html">10. beta_PCA.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_jitter_plot.html">11. beta_jitter_plot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_m_conversion.html">12. beta_m_conversion.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_profile_gene_centered.html">13. beta_profile_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_profile_region.html">14. beta_profile_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_stacked_barplot.html">15. beta_stacked_barplot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_stats.html">16. beta_stats.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_tSNE.html">17. beta_tSNE.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_topN.html">18. beta_topN.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_trichotmize.html">19. beta_trichotmize.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_Bayes.html">20. dmc_Bayes.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_bb.html">21. dmc_bb.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_fisher.html">22. dmc_fisher.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_glm.html">23. dmc_glm.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_logit.html">24. dmc_logit.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_nonparametric.html">25. dmc_nonparametric.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmc_ttest.html">26. dmc_ttest.py</a></li>
</ul>
<p class="caption"><span class="caption-text">Compare Differential Analysis Tools</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../compare_diff_tools.html">1. Compare Differential CpG Analysis Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../compare_diff_tools.html#p-value-distributions">2. P-value distributions</a></li>
</ul>
<p class="caption"><span class="caption-text">Liscense</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../liscense.html">LICENSE</a></li>
</ul>
<p class="caption"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reference.html">Reference</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CpGtools</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">2. </span>CpG_anno_position.py</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/demo/CpG_anno_position.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="cpg-anno-position-py">
<h1><span class="section-number">2. </span>CpG_anno_position.py<a class="headerlink" href="#cpg-anno-position-py" title="Permalink to this headline">¶</a></h1>
<div class="section" id="description">
<h2><span class="section-number">2.1. </span>Description<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h2>
<p>This program adds annotation information to each CpG based on its <strong>genomic position</strong>.</p>
</div>
<div class="section" id="notes">
<h2><span class="section-number">2.2. </span>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Input CpG (-i) and annotation (-a) BED files must have at least three columns, and must based on the same genome assembly   version</p></li>
<li><p>If multiple regions from the annotation BED file are overlapped with the <strong>same</strong>
CpG site, their names will be concatenated together.</p></li>
<li><p>Since the input (-i) is a regular BED foramt file, this module can be uesd to annotate any genomic regions of interest.</p></li>
</ul>
</div>
<div class="section" id="pre-computed-datasets">
<h2><span class="section-number">2.3. </span>Pre-computed datasets<a class="headerlink" href="#pre-computed-datasets" title="Permalink to this headline">¶</a></h2>
<dl class="simple">
<dt><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/hg19_ENCODE_338TF_130Cell_E3.bed.gz">hg19_ENCODE_338TF_130Cell_E3.bed.gz</a> (File size  = 108.2 MB)</dt><dd><p>Transcription factor (TF) binding sites identified from <a class="reference external" href="https://en.wikipedia.org/wiki/ChIP-sequencing">ChIP-seq</a> experiments performed
by the <a class="reference external" href="https://www.encodeproject.org/">ENCODE</a> project. Peaks from 1264 experiments
representing 338 transcription factors in 130 cell types are combined (N = 10,560,472). BED format file was
downloaded from the <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgTables">UCSC Tabel Browser</a>.</p>
</dd>
<dt><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/hg19_ENCODE_DNaseI_125Cells_V3.bed.gz">hg19_ENCODE_DNaseI_125Cells_V3.bed.gz</a> (File size  = 24.3 MB)</dt><dd><p><a class="reference external" href="https://en.wikipedia.org/wiki/DNase_I_hypersensitive_site">DNase I hypersensitivity sites</a> identified from
<a class="reference external" href="https://www.encodeproject.org/">ENCODE</a> <a class="reference external" href="https://en.wikipedia.org/wiki/DNase-Seq">DNase-seq</a> experiments. Peaks
from 125 cell types are combined (N = 1,867,665). BED format file was downloaded from
the <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgTables">UCSC Tabel Browser</a>.</p>
</dd>
<dt><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/hg19_ENCODE_chromHMM_states_9Cells.merge.bed.gz">hg19_ENCODE_chromHMM_states_9Cells.merge.bed.gz</a> (File size  = 32.7 MB)</dt><dd><p>Chromatin State Segmentation by <a class="reference external" href="https://www.nature.com/articles/nmeth.1906">chromHMM</a>
from <a class="reference external" href="https://www.encodeproject.org/">ENCODE</a>. Chromatin states across 9 cell
types (GM12878, H1-hESC, K562, HepG2, HUVEC, HMEC, HSMM, NHEK, NHLF) were learned by
integrating 9 factors (CTCF, H3K27ac, H3K27me3, H3K36me3, H3K4me1,
H3K4me2, H3K4me3, H3K9ac, H4K20me1 ) plus input. A total of 15 states were identified,
include: State-1 (Active Promoter), state-2 (Weak Promoter), state-3 (Inactive/poised
Promoter), state-4 and 5 (Strong enhancer), state-6 and 7 (Weak/poised enhancer),
state-8 (insulator), state-9 (Transcriptional transition), state-10 (Transcriptional
elongation), state-11 (Weak transcribed), state-12 (Polycomb-repressed), state-13
(Heterochromatin or low signal), state-14 and 15 (Repetitive/Copy Number Variation).
The Original chromatin state BED file was downloaded from the <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgTables">UCSC Tabel Browser</a>.</p>
</dd>
<dt><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/hg19_FANTOM_enhancers_phase_1_and_2.bed.gz">hg19_FANTOM_enhancers_phase_1_and_2.bed.gz</a></dt><dd><p><a class="reference external" href="http://fantom.gsc.riken.jp/5/">PHANTOM5</a> human permissive enhancers downloaded from <a class="reference external" href="http://fantom.gsc.riken.jp/5/datafiles/latest/extra/Enhancers/human_permissive_enhancers_phase_1_and_2_expression_tpm_matrix.txt.gz">here</a>.</p>
</dd>
<dt><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/hg19_ENCODE_H3K4me1_11_cellLines_ChIP.bed.gz">hg19_ENCODE_H3K4me1_11_cellLines_ChIP.bed.gz</a> (File size  = 12.2 MB)</dt><dd><p>H3K4me1 (marker of active and primed enhancer) peaks identified from <a class="reference external" href="https://www.encodeproject.org/">ENCODE</a> histone ChIP-seq experiments. Peaks from 11 cell
types (GM12878, H1-hESC, HMEC, HSMM, HUVEC, HeLaS3, HepG2, K562, Monocytes-CD14+_RO01746,
NHEK, NHLF) are combined (N = 1,435,550)</p>
</dd>
<dt><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/hg19_ENCODE_H3K4me3_11_cellLines_ChIP.bed.gz">hg19_ENCODE_H3K4me3_11_cellLines_ChIP.bed.gz</a> (File size  = 4.5 MB)</dt><dd><p>H3K4me3 (marker of promoter) peaks identified from <a class="reference external" href="https://www.encodeproject.org/">ENCODE</a> histone ChIP-seq experiments. Peaks from 11 cell
types (GM12878, H1-hESC, HMEC, HSMM, HUVEC, HeLaS3, HepG2, K562, Monocytes-CD14+_RO01746,
NHEK, NHLF) are combined (N = 525,824)</p>
</dd>
<dt><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/hg19_ENCODE_H3K27ac_11_cellLines_ChIP.bed.gz">hg19_ENCODE_H3K27ac_11_cellLines_ChIP.bed.gz</a> (File size  = 5.7 MB)</dt><dd><p>H3K27ac (marker of active enhancer) peaks identified from <a class="reference external" href="https://www.encodeproject.org/">ENCODE</a> histone ChIP-seq experiments. Peaks from 11 cell
types (GM12878, H1-hESC, HMEC, HSMM, HUVEC, HeLaS3, HepG2, K562, Monocytes-CD14+_RO01746,
NHEK, NHLF) are combined (N = 665,650)</p>
</dd>
</dl>
<p>These BED files were lifted over to hg38/GRCh38 using <a class="reference external" href="http://crossmap.sourceforge.net/">CrossMap</a>.
The hg38-based annotation files are available from <a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/">here</a></p>
</div>
<div class="section" id="options">
<h2><span class="section-number">2.4. </span>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-i <var>INPUT_FILE</var></span>, <span class="option">--input_file=<var>INPUT_FILE</var></span></kbd></dt>
<dd><p>Input CpG file in BED3+ format.</p>
</dd>
<dt><kbd><span class="option">-a <var>ANNO_FILE</var></span>, <span class="option">--annotation=<var>ANNO_FILE</var></span></kbd></dt>
<dd><p>Input annotation file in BED3+ format.</p>
</dd>
<dt><kbd><span class="option">-w <var>WINDOW_SIZE</var></span>, <span class="option">--window=<var>WINDOW_SIZE</var></span></kbd></dt>
<dd><p>Size of window centering on the middle-point of each
genomic region defined in the annotation BED file
(i.e., window_size*0.5 will be extended to up- and
down-stream from the middle point of each genomic
region). default=100</p>
</dd>
<dt><kbd><span class="option">-o <var>OUT_FILE</var></span>, <span class="option">--output=<var>OUT_FILE</var></span></kbd></dt>
<dd><p>The prefix of the output file.</p>
</dd>
<dt><kbd><span class="option">-l</span>, <span class="option">--header</span></kbd></dt>
<dd><p>If True, the first row of input CpG file is header.
default=False</p>
</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="input-files-examples">
<h2><span class="section-number">2.5. </span>Input files (examples)<a class="headerlink" href="#input-files-examples" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/test/test_01.hg19.bed6">test_01.hg19.bed6</a></p></li>
<li><p><a class="reference external" href="https://sourceforge.net/projects/cpgtools/files/data/hg19_ENCODE_338TF_130Cell_E3.bed.gz/download">hg19_ENCODE_338TF_130Cell_E3.bed.gz</a></p></li>
</ul>
</div>
<div class="section" id="command">
<h2><span class="section-number">2.6. </span>Command<a class="headerlink" href="#command" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$CpG_anno_position.py  -l -a hg19_ENCODE_338TF_130Cell_E3.bed.gz -i test_01.hg19.bed6 -o output
</pre></div>
</div>
</div>
<div class="section" id="output-files">
<h2><span class="section-number">2.7. </span>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>output.anno.txt</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head -5 output.anno.txt
#Chrom Start   End     Name    Beta    Strand  hg19_ENCODE_338TF_130Cell_E3.bed
chr1   10847   10848   cg26928153      0.8965  +       N/A
chr1   10849   10850   cg16269199      0.7915  +       N/A
chr1   15864   15865   cg13869341      0.9325  +       N/A
chr1   534241  534242  cg24669183      0.7941  +       FOXA2,MNT
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="CpG_anno_probe.html" class="btn btn-neutral float-right" title="3. CpG_anno_probe.py" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="CpG_aggregation.html" class="btn btn-neutral float-left" title="1. CpG_aggregation.py" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2019, Liguo Wang
      <span class="lastupdated">
        Last updated on July 7, 2019.
      </span>

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>