

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>1. Input file and data format &mdash; CpGtools documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=c97855ec" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="1. Pre-compiled datasets" href="dataset.html" />
    <link rel="prev" title="3. CpGtools Release history" href="change_log.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >

          
          
          <a href="index.html" class="icon icon-home">
            CpGtools
              <img src="_static/CpGtools_logo3.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">CpGtools Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview.html">1. Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">2. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="change_log.html">3. CpGtools Release history</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Data Format</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">1. Input file and data format</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#bed-file">1.1. BED file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#proportion-values">1.2. Proportion values</a></li>
<li class="toctree-l2"><a class="reference internal" href="#beta-values">1.3. Beta values</a></li>
<li class="toctree-l2"><a class="reference internal" href="#m-values">1.4. M values</a></li>
<li class="toctree-l2"><a class="reference internal" href="#convert-beta-value-to-m-value-or-vice-versa">1.5. Convert Beta value to M value or <em>vice versa</em></a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Test and Demo Datasets</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="dataset.html">1. Pre-compiled datasets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Information</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_aggregation.html">1. CpG_aggregation.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_anno_position.html">2. CpG_anno_position.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_anno_probe.html">3. CpG_anno_probe.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_density_gene_centered.html">4. CpG_density_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_distrb_chrom.html">5. CpG_distrb_chrom.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_distrb_gene_centered.html">6. CpG_distrb_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_distrb_region.html">7. CpG_distrb_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_logo.html">8. CpG_logo.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_to_gene.html">9. CpG_to_gene.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_PCA.html">10. beta_PCA.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_UMAP.html">11. beta_UMAP.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_jitter_plot.html">12. beta_jitter_plot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_m_conversion.html">13. beta_m_conversion.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_profile_gene_centered.html">14. beta_profile_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_profile_region.html">15. beta_profile_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_remove_batch_effects.html">16. beta_remove_batch_effects.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_selectNBest.html">17. beta_selectNBest.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_stacked_barplot.html">18. beta_stacked_barplot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_stats.html">19. beta_stats.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_tSNE.html">20. beta_tSNE.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_topN.html">21. beta_topN.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_trichotmize.html">22. beta_trichotmize.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_Bayes.html">23. dmc_Bayes.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_bb.html">24. dmc_bb.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_fisher.html">25. dmc_fisher.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_glm.html">26. dmc_glm.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_logit.html">27. dmc_logit.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_nonparametric.html">28. dmc_nonparametric.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_ttest.html">29. dmc_ttest.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/predict_sex.html">30. predict_sex.py</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Compare Differential Analysis Tools</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="compare_diff_tools.html">1. Compare Differential CpG Analysis Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_diff_tools.html#p-value-distributions">2. P-value distributions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Liscense</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="liscense.html">LICENSE</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="reference.html">Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: white" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CpGtools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">1. </span>Input file and data format</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/input_format.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="input-file-and-data-format">
<h1><span class="section-number">1. </span>Input file and data format<a class="headerlink" href="#input-file-and-data-format" title="Link to this heading"></a></h1>
<section id="bed-file">
<h2><span class="section-number">1.1. </span>BED file<a class="headerlink" href="#bed-file" title="Link to this heading"></a></h2>
<p>BED (Browser Extensible Data) format is commonly used to describe blocks of genome. The
BED format consists of one line per feature, each containing 3-12 columns of data. It is
0-based (meaning the first base of a chromosome is numbered 0). It is s left-open,
right-closed. For example, the bed entry “chr1 10 15” contains the 11-th, 12-th, 13-th,
14-th and 15-th bases of chromosome-1.</p>
<dl class="simple">
<dt>BED12 file</dt><dd><p>The standard BED file which has 12 fields. Each row in this file describes a
gene or an array of disconnected genomic regions. Details are described <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">here</a></p>
</dd>
<dt>BED3 file</dt><dd><p>Only has the first three required fields (chrom, chromStart, chromEnd). Each row is
used to represent a single genomic region where “score” and “strand” are not necessary.</p>
</dd>
<dt>BED3+ file</dt><dd><p>Has at least three columns (chrom, chromStart, chromEnd). It could have other
columns, but these additional columns will be ignored.</p>
</dd>
<dt>BED6 file</dt><dd><p>Has the first six fields (chrom, chromStart, chromEnd, name, score, strand). Each row
is used to represent a single genomic region and their associated scores, or in cases
where “strand” information is essential.</p>
</dd>
<dt>BED6+ file</dt><dd><p>Has at least six columns (chrom, chromStart, chromEnd, name, score, stand). It could
have other columns, but these additional columns will be ignored.</p>
</dd>
</dl>
</section>
<section id="proportion-values">
<h2><span class="section-number">1.2. </span>Proportion values<a class="headerlink" href="#proportion-values" title="Link to this heading"></a></h2>
<p>In <a class="reference external" href="https://en.wikipedia.org/wiki/Bisulfite_sequencing">bisulfite sequencing</a>
(RRBS or WGBS), the methylation level of a particular CpG or
region can be represented by a “proportion” value. We define the proportion value as a
pair of integers separated by comma (“,”) with the first integer (m, 0 &lt;- m &lt;- n)
representing “number of methylated reads” and the second integer (n, n &gt;- 0) representing
“number of total reads”. for example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span>   <span class="mi">1</span><span class="p">,</span><span class="mi">27</span>    <span class="mi">2</span><span class="p">,</span><span class="mi">159</span>   <span class="c1">#Three proportions values indicated 3 hypo-methylated loci</span>
<span class="mi">7</span><span class="p">,</span><span class="mi">7</span>    <span class="mi">17</span><span class="p">,</span><span class="mi">19</span>   <span class="mi">30</span><span class="p">,</span><span class="mi">34</span>   <span class="c1">#Three proportions values indicated 3 hyper-methylated loci</span>
</pre></div>
</div>
</section>
<section id="beta-values">
<h2><span class="section-number">1.3. </span>Beta values<a class="headerlink" href="#beta-values" title="Link to this heading"></a></h2>
<p>The Beta-value is a value between 0 and 1, which can be interpreted as the approximation
of the percentage of methylation for a given CpG or locus. One can convert proportion
value into beta value, but not vice versa. In the equation below, C is the “probe intensity”
or “read count” of methylated allele, while U is the “probe intensity” or “read count” of
unmethylated allele.</p>
<div class="math notranslate nohighlight">
\[\beta=\frac{C}{U+C}, (0 \leq \beta \leq 1)\]</div>
</section>
<section id="m-values">
<h2><span class="section-number">1.4. </span>M values<a class="headerlink" href="#m-values" title="Link to this heading"></a></h2>
<p>The M-value is calculated as the log2 ratio of the probe intensities (or read counts) of
methylated allele versus unmethylated allele. In the equation below, C is the “probe
intensity” or “read count” of methylated allele, while U is the “probe intensity” or
“read count” of unmethylated allele. w is the offset or pseudo count added to both
denominator and numerator to avoid unexpected big changes and performing log
transformation on zeros.</p>
<div class="math notranslate nohighlight">
\[M=\log _{2}\left(\frac{C+w}{U+w}\right)\]</div>
</section>
<section id="convert-beta-value-to-m-value-or-vice-versa">
<h2><span class="section-number">1.5. </span>Convert Beta value to M value or <em>vice versa</em><a class="headerlink" href="#convert-beta-value-to-m-value-or-vice-versa" title="Link to this heading"></a></h2>
<p>The relationship between Beta-value and M-value is shown as equation and figure:</p>
<div class="math notranslate nohighlight">
\[\beta=\frac{2^{M}}{2^{M}+1} ; M=\log _{2}\left(\frac{\beta}{1-\beta}\right)\]</div>
<a class="reference internal image-reference" href="_images/beta_vs_M_curve.png"><img alt="_images/beta_vs_M_curve.png" class="align-center" src="_images/beta_vs_M_curve.png" style="width: 320.0px; height: 320.0px;" />
</a>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="change_log.html" class="btn btn-neutral float-left" title="3. CpGtools Release history" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="dataset.html" class="btn btn-neutral float-right" title="1. Pre-compiled datasets" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Liguo Wang.
      <span class="lastupdated">Last updated on Oct 07, 2025.
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>