

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>1. Overview &mdash; CpGtools documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=c97855ec" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2. Installation" href="installation.html" />
    <link rel="prev" title="&lt;no title&gt;" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >

          
          
          <a href="index.html" class="icon icon-home">
            CpGtools
              <img src="_static/CpGtools_logo3.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">CpGtools Documentation</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">1. Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#cpg-position-analysis-modules">1.1. CpG position analysis modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#cpg-signal-analysis-modules">1.2. CpG signal analysis modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#differential-cpg-analysis-modules">1.3. Differential CpG analysis modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#predictive-modules">1.4. Predictive modules</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">2. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="change_log.html">3. CpGtools Release history</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Data Format</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="input_format.html">1. Input file and data format</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Test and Demo Datasets</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="dataset.html">1. Pre-compiled datasets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Information</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_aggregation.html">1. CpG_aggregation.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_anno_position.html">2. CpG_anno_position.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_anno_probe.html">3. CpG_anno_probe.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_density_gene_centered.html">4. CpG_density_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_distrb_chrom.html">5. CpG_distrb_chrom.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_distrb_gene_centered.html">6. CpG_distrb_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_distrb_region.html">7. CpG_distrb_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_logo.html">8. CpG_logo.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/CpG_to_gene.html">9. CpG_to_gene.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_PCA.html">10. beta_PCA.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_UMAP.html">11. beta_UMAP.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_jitter_plot.html">12. beta_jitter_plot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_m_conversion.html">13. beta_m_conversion.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_profile_gene_centered.html">14. beta_profile_gene_centered.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_profile_region.html">15. beta_profile_region.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_remove_batch_effects.html">16. beta_remove_batch_effects.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_selectNBest.html">17. beta_selectNBest.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_stacked_barplot.html">18. beta_stacked_barplot.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_stats.html">19. beta_stats.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_tSNE.html">20. beta_tSNE.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_topN.html">21. beta_topN.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/beta_trichotmize.html">22. beta_trichotmize.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_Bayes.html">23. dmc_Bayes.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_bb.html">24. dmc_bb.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_fisher.html">25. dmc_fisher.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_glm.html">26. dmc_glm.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_logit.html">27. dmc_logit.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_nonparametric.html">28. dmc_nonparametric.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/dmc_ttest.html">29. dmc_ttest.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/predict_sex.html">30. predict_sex.py</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Compare Differential Analysis Tools</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="compare_diff_tools.html">1. Compare Differential CpG Analysis Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_diff_tools.html#p-value-distributions">2. P-value distributions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Liscense</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="liscense.html">LICENSE</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="reference.html">Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: white" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CpGtools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">1. </span>Overview</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/overview.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="overview">
<h1><span class="section-number">1. </span>Overview<a class="headerlink" href="#overview" title="Link to this heading"></a></h1>
<p>CpGtools package provides a number of Python programs to annotate, QC, visualize, and
analyze DNA methylation data generated from Illumina
<a class="reference external" href="https://support.illumina.com/array/array_kits/infinium_humanmethylation450_beadchip_kit.html">HumanMethylation450 BeadChip (450K)</a>,
<cite>MethylationEPIC BeadChip (850K) &lt;https://www.illumina.com/documents/products/datasheets/datasheet_CytoSNP850K_POP.pdf&gt;</cite>, <cite>930K &lt;https://www.illumina.com/products/by-type/microarray-kits/infinium-methylation-epic.html&gt;</cite> or
<a class="reference external" href="https://www.illumina.com/science/sequencing-method-explorer/kits-and-arrays/rrbs-seq-scrrbs.html">RRBS / WGBS</a>.</p>
<p>These programs can be divided into four classes:</p>
<ul class="simple">
<li><p>CpG position analysis modules</p></li>
<li><p>CpG signal analysis modules</p></li>
<li><p>Differential CpG analysis modules</p></li>
<li><p>Predictive modules (under development)</p></li>
</ul>
<section id="cpg-position-analysis-modules">
<h2><span class="section-number">1.1. </span>CpG position analysis modules<a class="headerlink" href="#cpg-position-analysis-modules" title="Link to this heading"></a></h2>
<p>These modules are primarily used to analyze CpG’s genomic locations.</p>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>Name</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_aggregation.html">CpG_aggregation.py</a></p></td>
<td><p>Aggregate proportion values of CpGs that located in give genomic regions (eg. CpG islands, promoters, exons, etc.).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_anno_position.html">CpG_anno_position.py</a></p></td>
<td><p>Add annotation information CpGs according to their genomic coordinates.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_anno_probe.html">CpG_anno_probe.py</a></p></td>
<td><p>Add annotation information to 450K/850K probes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_density_gene_centered.html">CpG_density_gene_centered.py</a></p></td>
<td><p>Generate the <a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/_images/CpG_density.png">CpG density (count) profile</a> over gene body and the up/down-stream intergenic regions.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_distrb_chrom.html">CpG_distrb_chrom.py</a></p></td>
<td><p>Calculate the distribution of CpG over chromosomes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_distrb_gene_centered.html">CpG_distrb_gene_centered.py</a></p></td>
<td><p>Calculate the distribution of CpG over <a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/_images/geneDist.png">gene-centered genomic regions</a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_distrb_region.html">CpG_distrb_region.py</a></p></td>
<td><p>Calculate the distribution of CpG over <a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/_images/regionDist.png">user-specified genomic regions</a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_logo.html">CpG_logo.py</a></p></td>
<td><p>Generate a <a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/_images/450_CH.logo.png">DNA motif logo</a> and matrices for a given set of CpGs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/CpG_to_gene.html">CpG_to_gene.py</a></p></td>
<td><p>Assign CpGs to their putative target genes. It uses the algorithm similar to <a class="reference external" href="http://great.stanford.edu/public/html/">GREAT</a>.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="cpg-signal-analysis-modules">
<h2><span class="section-number">1.2. </span>CpG signal analysis modules<a class="headerlink" href="#cpg-signal-analysis-modules" title="Link to this heading"></a></h2>
<p>These modules are primarily used to analyze CpG’s DNA methylation beta values</p>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>Name</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_PCA.html">beta_PCA.py</a></p></td>
<td><p>Perform <a class="reference external" href="https://en.wikipedia.org/wiki/Principal_component_analysis">PCA</a> (principal component analysis) for samples.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_jitter_plot.html">beta_jitter_plot.py</a></p></td>
<td><p>Generate <a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/_images/Jitter.png">jitter plot</a> (a.k.a. strip chart) and bean plot for each sample.”</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_m_conversion.html">beta_m_conversion.py</a></p></td>
<td><p>Convert Beta-value into M-value or <em>vice versa</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_profile_gene_centered.html">beta_profile_gene_centered.py</a></p></td>
<td><p>Calculate the <a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/_images/gene_profile.png">methylation profile</a> (i.e., average beta value) for genomic regions around genes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_profile_region.html">beta_profile_region.py</a></p></td>
<td><p>Calculate <a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/_images/region_profile.png">methylation profile</a> (i.e. average beta value) around the user-specified genomic regions.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_stacked_barplot.html">beta_stacked_barplot.py</a></p></td>
<td><p>Create <a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/_images/stacked_bar.png">stacked barplot</a> for each sample. The stacked barplot showing the proportions of CpGs whose beta values are falling into [0,0.25], [0.25,0.5], [0.5,0.75],[0.75,1]</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_stats.html">beta_stats.py</a></p></td>
<td><p>Summarize basic information on CpGs located in each genomic region.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_tSNE.html">beta_tSNE.py</a></p></td>
<td><p>Perform <a class="reference external" href="https://lvdmaaten.github.io/tsne/">t-SNE</a> (t-Distributed Stochastic Neighbor Embedding) analysis for samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_topN.html">beta_topN.py</a></p></td>
<td><p>Select the top N most variable CpGs (according to standard deviation) from the input file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_trichotmize.html">beta_trichotmize.py</a></p></td>
<td><p>Use <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.mixture.BayesianGaussianMixture.html">Bayesian Gaussian Mixture model</a> to trichotmize beta values into three status: ‘Un-methylated’,’Semi-methylated’, ‘Full-methylated’, and ‘unassigned’.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_UMAP.html">beta_UMAP.py</a></p></td>
<td><p>Perform <a class="reference external" href="https://github.com/lmcinnes/umap">UMAP</a> (Uniform Manifold Approximation and Projection) for samples.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_selectNBest.html">beta_selectNbest.py</a></p></td>
<td><p>Select the K best features using ANOVA, Mutual information or Chi-squared stat.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/beta_remove_batch_effects.html">beta_remove_batch_effects.py</a></p></td>
<td><p>Corrects batch effect using the <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/16632515/">combat</a> algorithm.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="differential-cpg-analysis-modules">
<h2><span class="section-number">1.3. </span>Differential CpG analysis modules<a class="headerlink" href="#differential-cpg-analysis-modules" title="Link to this heading"></a></h2>
<p>These modules are primarily used to identify CpGs that are differentially methylated between groups</p>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>Name</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/dmc_Bayes.html">dmc_Bayes.py</a></p></td>
<td><p>Differential CpG analysis using the Bayesian approach. (for 450K/850K data)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/dmc_bb.html">dmc_bb.py</a></p></td>
<td><p>Differential CpG analysis using the beta-binomial model. (for RRBS/WGBS count data)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/dmc_fisher.html">dmc_fisher.py</a></p></td>
<td><p>Differential CpG analysis using Fisher’s Exact Test. (for RRBS/WGBS count data)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/dmc_glm.html">dmc_glm.py</a></p></td>
<td><p>Differential CpG analysis using the <a class="reference external" href="https://en.wikipedia.org/wiki/Generalized_linear_model">GLM</a> generalized liner model. (for 450K/850K data)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/dmc_logit.html">dmc_logit.py</a></p></td>
<td><p>Differential CpG analysis using logistic regression model. (for RRBS/WGBS count data)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/dmc_nonparametric.html">dmc_nonparametric.py</a></p></td>
<td><p>Differential CpG analysis using <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html">Mann-Whitney U test</a> for two group comparison, and the <a class="reference external" href="https://en.wikipedia.org/wiki/Kruskal%E2%80%93Wallis_one-way_analysis_of_variance">Kruskal-Wallis H-test</a> for multiple groups comparison.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/dmc_ttest.html">dmc_ttest.py</a></p></td>
<td><p>Differential CpG analysis using T test. (for 450K/850K data)</p></td>
</tr>
</tbody>
</table>
</section>
<section id="predictive-modules">
<h2><span class="section-number">1.4. </span>Predictive modules<a class="headerlink" href="#predictive-modules" title="Link to this heading"></a></h2>
<p>These modules are primarily used to predict phenotypes from DNA methylation data</p>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>Name</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://cpgtools.readthedocs.io/en/latest/demo/predict_sex.html">predict_sex.py</a></p></td>
<td><p>Predict sex based on the semi-methylation (also known as genomic imprinting) ratio.</p></td>
</tr>
</tbody>
</table>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="&lt;no title&gt;" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="installation.html" class="btn btn-neutral float-right" title="2. Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Liguo Wang.
      <span class="lastupdated">Last updated on Oct 07, 2025.
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>